docs/madr/decisions/020-kds-redesign.md

```diff
@@ -0,0 +1,238 @@
+
+```proto
+service GlobalKDSService {
+ ...
+ rpc ZoneToGlobalStream(stream envoy.service.discovery.v3.DeltaDiscoveryRequest)
+     returns (stream envoy.service.discovery.v3.DeltaDiscoveryResponse);
+
+ rpc GlobeToZoneStream(stream envoy.service.discovery.v3.DeltaDiscoveryRequest)
+     returns (stream envoy.service.discovery.v3.DeltaDiscoveryRequest);
+}
+
+```
+
+In the first iteration, we can introduce only `ZoneToGlobalStream` with delta xDS, and in the future, we could introduce `GlobeToZoneStream` and also use delta xDS. In the code, I've noticed that we are using our copy of the Snapshot cache and the code is a bit outdated. As a part of that task, we should check if we can change to use `go-control-plane` cache implementation or copy and update our current implementation.
+
+### Handshake
+
+Is it possible for the zone control-plane to be newer than the global control-plane?
+
+If the answer is no, then a handshake is not necessary as the client will use the latest available version.
+
+If the answer is yes, then a handshake API may need to be implemented. The GRPC Reflection API allows services to be discovered on the server side, but not on the client side.
+
+The output of the reflection functionality is:
+
+```
+["grpc.reflection.v1alpha.ServerReflection", "kuma.mesh.v1alpha1.GlobalKDSService", "kuma.mesh.v1alpha1.MultiplexService"]```
+```
+
+How would a handshake work?
+
+1. The server exposes the RPC `handshake`.
+2. The client initiates a connection to the server.
+3. The client sends a request that includes the Kuma version and the supported KDS versions:
+```
+ kumaVersion: "2.1.0",
+ supportedKdsVersions: ["0.1.0", "0.2.0"],
+```
+4. Server checks the request and responds with the newest version working for both
+```
+ kdsVersion: "0.2.0"
+```
+5. The client needs to create RPC clients for the specific version.
+
+### Multitenant
+
+To link multitenant control-planes with a global one, it's necessary to establish a method of recognition. One approach to accomplish this is by extracting tenant specifics from an authentication token. These details can then be transmitted along with a request to indicate the specific tenant and retrieve solely their deployment information. Another option is to supply this information via a header. It appears that accommodating multitenancy doesn't require additional exertion.
+
+### Create resources on the zone
+
+We intend to introduce a new feature that enables the creation of resources at the zone level. These resources will be exclusively applicable to the specific zone, and we will synchronize them to the global level as read-only. To distinguish the zone resources from global ones, we will prefix them with the name of the zone.
+
+```
+zone-1.example-timeout
+```
```

Comment:
I feel like this feature needs it's own separate MADR.

I was thinking that when a CP is receiving a policy from a zone it modifies the top level targetRef to add `kuma.io/zone`.

The zone name being prefixed is I think something pretty standard for all synced entities right?

Reply:
> What's the benefit of letting users create policies in Zone CP?

Not giving write access for teams to Global CP. They could just apply resources on their cluster + namespace.

Reply:
I agree with Jakub that it would be nice to make it at the zone and apply there. Added note to cover implementation in a separate MADR

Reply:
> What's the benefit of letting users create policies in Zone CP?

In the event of a global CP outage people can still do things in their zones without hackery like removing the webhook.